0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+This is certainly a well-formed and sound construct, but far less powerful than the proposed version, especially
+when sequence comprehensions are excluded. It also still runs into some of the same semantic challenges around
+ordering if `C` is allowed to be a set or multiset anyway, since verification of such a loop still needs to be aware of
+this ordering. Offering `foreach` loops along with sequence comprehensions means we can recommend the latter where possible,
+with the former as a fallback that is inevitably more complicated to verify.
+
+# Prior art
+[prior-art]: #prior-art
+
+Sequence comprehensions bear a strong resemblance to list comprehensions in Python, which also include similar abilities to 
+bind multiple variables and filter the enumerated values: `[(x, y) for x in [1,2,3] for y in [3,1,4] if x != y]`
+
+Most mainstream programming languages include rich libraries for working with collections and iterators,
+and define much the same operations such as "filter" and "flatMap".
+The proposed Dafny features provide similar functionality at a higher level
+of abstraction more amenable to verification: quantification expressions are used to succinctly declare the behavior of
+an operation that in spirit applies several such operations. See the "Future Possibilities" section for ideas for
+pushing this philosophy even further.
+
+The [JMatch language](https://www.cs.cornell.edu/andru/papers/padl03.pdf), an extension of Java to support
+pattern matching, includes a `foreach` loop remarkably similar to the proposal here: it is parameterized by
+a boolean expression and iterates through all matching bindings. For example, the statement `foreach(a[int i] != 0) n += i;`
+sums the indices of all non-zero elements of the array `a`. The JMatch version declares the bound variables inline in
+the boolean expression, which is also proposed here as a future possibility below.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+One large open question is how difficult it will be to convince the verifier the following is correct,
+since it will be a common idiom for depending on the unpredictable enumeration ordering of a set:
+
+```dafny
+function method SortedElements(s: set<int>): seq<int> 
+  ensures multiset(SortedElements(s)) == multiset(s)
+  ensures Seq.IsSorted(SortedElements(s), (a, b) => a < b)
+{
+  // Assume Seq.Sorted and Seq.IsSorted are part of the standard library
+  Seq.Sorted(seq x | x in s, (a, b) => a < b)
+}
+
+method Foo() {
+  var smallSet := {1, 2};
+  var smallSortedSeq := SortedElements(smallSet);
+  // smallSortedSeq is fully specified, but can the verifier figure that out?
+  assert smallSortedSeq == [1, 2];
+}
+```
+
+One of the first milestones when implementing this RFC will be to prototype just enough verification
+plumbing to evaluate how difficult proving such properties will be, and how many additional axioms
+may need to be added to the Dafny prelude.
```

Comment:
Here's a proof with Dafny as of today:

```dafny
predicate method IsLower(s: set<int>, m: int) {
  forall y :: y in s ==> m <= y
}

function Min(s: set<int>) : (m: int)
  requires s != {}
  ensures m in s
  ensures IsLower(s, m)
{
  var m :| m in s;
  if y :| y in s && y < m then
    var s' := s - {m};
    assert s == s' + {m};
    Min(s')
  else m
}

predicate Increasing(s: seq<int>) {
  forall i, j | 0 <= i < j < |s| :: s[i] <= s[j]
}

lemma IncreasingConcat(hd: int, tl: seq<int>)
  requires Increasing(tl)
  requires forall x | x in tl :: hd <= x
  ensures Increasing([hd] + tl)
{
  var s := [hd] + tl;
  forall i, j | 0 <= i < j < |s| ensures s[i] <= s[j] {
    assert s[j] in tl;
  }
}

lemma IsLowerSeq(s: set<int>, sq: seq<int>, m: int)
  requires IsLower(s, m)
  requires multiset(sq) == multiset(s)
  ensures forall x | x in sq :: m <= x
{
  forall x | x in sq ensures m <= x {
    assert x in multiset(sq);
  }
}

function method Sort(sq: seq<int>) : (sq': seq<int>)
  ensures Increasing(sq')
  ensures multiset(sq') == multiset(sq)

lemma MultiSetAddInj<T>(b: multiset<T>, a1: multiset<T>, a2: multiset<T>)
  requires b + a1 == b + a2
  ensures a1 == a2
{
  assert forall k :: a1[k] == a2[k] by {
    forall k {
      calc {
        a1[k];
        (b[k] + a1[k]) - b[k];
        (b + a1)[k] - b[k];
        (b + a2)[k] - b[k];
        (b[k] + a2[k]) - b[k];
        a2[k];
      }
    }
  }
}

lemma SortedElements_Unique(s: set<int>, sq: seq<int>)
  requires multiset(sq) == multiset(s)
  requires Increasing(sq)
  ensures sq == SortedElements(s)
{
  if sq == [] {
    assert |sq| == |multiset(s)| == 0;
  } else {
    var hd := sq[0];
    var sq' := sq[1..];
    var s' := s - {hd};
    calc { hd in sq; hd in multiset(sq); hd in multiset(s); hd in s; }

    assert sq == [hd] + sq';
    assert multiset(sq) == multiset([hd]) + multiset(sq');
    assert s == {hd} + s';
    assert multiset(sq) == multiset([hd]) + multiset(s');
    MultiSetAddInj(multiset([hd]), multiset(sq'), multiset(s'));

    assert Increasing(sq');
    SortedElements_Unique(s', sq');
  }
}

function SortedElements(s: set<int>): (sq: seq<int>)
  ensures multiset(sq) == multiset(s)
  ensures Increasing(sq)
{
  if s == {} then []
  else
    var m := Min(s);
    var s := s - {m};
    var sq := SortedElements(s);
    IsLowerSeq(s, sq, m); IncreasingConcat(m, sq);
    [m] + sq
} by method {
  sq := [];
  var leftover := s;
  ghost var removed := {};
  while leftover != {}
    decreases leftover
    invariant leftover !! removed
    invariant multiset(sq) == multiset(removed)
    invariant multiset(leftover + removed) == multiset(s)
  {
    var x :| x in leftover;
    leftover := leftover - {x};
    removed := removed + {x};
    sq := sq + [x];
  }
  sq := Sort(sq);
  SortedElements_Unique(s, sq);
}

method Foo() {
  var smallSet := {1, 2};
  var smallSortedSeq := SortedElements(smallSet);
  // smallSortedSeq is fully specified, but can the verifier figure that out?
  assert multiset(smallSortedSeq) == multiset{1, 2};
  assert smallSortedSeq == [1, 2];
}
```

Can we think about how that proof would change?